messages/jsonschema/scripts/templates/perl.pm.erb

```diff
@@ -0,0 +1,137 @@
+package Cucumber::Messages;
+
+# DO NOT CHANGE THIS FILE!!
+
+# The code was auto-generated by this script:
+# https://github.com/cucumber/common/blob/main/messages/jsonschema/scripts/codegen.rb
+
+=head1 NAME
+
+=head1 SYNOPSIS
+
+=head1 DESCRIPTION
+
+
+=cut
+
+use strict;
+use warnings;
+
+use Cucumber::Messages::Message;
+
+=head1 MESSAGE CLASSES
+
+=cut
+
+<%- @schemas.sort.each do |key, schema| %>
+
+package Cucumber::Messages::<%= class_name(key) %> {
+
+=head2 Cucumber::Messages::<%= class_name(key) %>
+
+=head3 DESCRIPTION
+
+Represents the <%= class_name(key) %> message in Cucumber's
+L<message protocol|https://github.com/cucumber/common/tree/main/messages#readme>.
+
+<%= format_description(schema['description']) %>
+
+=head3 ATTRIBUTES
+
+=cut
+
+use Moo;
+extends 'Cucumber::Messages::Message';
+
+use Scalar::Util qw( blessed );
+
+my %types = (
+<%- schema['properties'].each do |property_name, property| -%>
+   <%= underscore(property_name) %> => '<%= type_for(class_name(key), property_name, property) %>',
+<%- end -%>
+);
+
+# This is a work-around for the fact that Moo doesn't have introspection
+# and Perl doesn't have boolean values...
+sub _types {
+    return \%types;
+}
+
+=pod
+around BUILDARGS => sub {
+   my ( $orig, $class, %args ) = @_;
+
+<%- schema['properties'].each do |property_name, property| -%>
+<%- if property['type'] == 'array'
+       items_ref = property.dig('items', '$ref')
+       if items_ref-%>
+    $args{<%= underscore(property_name) %>} = [
+       map { blessed( $_ ) ? $_ : Cucumber::Messages::<%= class_name(items_ref) %>->new( %$_ ) }
+       @{ $args{<%= underscore(property_name) %>} }
+       ]
+       if $args{<%= underscore(property_name) %>};
+
+<%-    end
+    end -%>
+<%- if property['$ref'] -%>
+    $args{<%= underscore(property_name) %>} =
+       Cucumber::Messages::<%= type_for(class_name(key), property_name, property) %>->new( %{ $args{<%= underscore(property_name) %>} } )
```

Comment:
This seems to generate things like the following:
```perl
Cucumber::Messages::Cucumber::Messages::Location->new()
```

Is this expected?

Reply:
No. Thansks for spotting that. The serializer/deserializer test doesn't capture this problem, so I'll need to implement more tests :-).

This error is a consequence of having changed the definition of `class_name()` and not working the consequences fully through. ETOOMUCHLATENIGHTCODING?

I'll fix in a sec.